### PR TITLE
Fix/video calls qtmegachatapi crashes

### PR DESCRIPTION
**Pull request by jgandres**

### CODE REVIEW

File: examples/qtmegachatapi/callGui.cpp [hunk 1]
```diff
@@ -1,70 +1,108 @@
...
-    drawAvatar(image, letter, client.myHandle());
+    int nPeers = mChatWindow->mChatRoom->getPeerCount();
+    megachat::MegaChatHandle peerHandle = megachat::MEGACHAT_INVALID_HANDLE;
+    const char *title = NULL;
+    for (int i = 0; i < nPeers; i++)
+    {
+        if (mChatWindow->mChatRoom->getPeerHandle(i) != mChatWindow->mMegaChatApi->getMyUserHandle())
+        {
+            peerHandle = mChatWindow->mChatRoom->getPeerHandle(i);
+            title = mChatWindow->mChatRoom->getPeerFullname(i);
```
Comment (sergiohs84) [inline examples/qtmegachatapi/callGui.cpp]: You take the ownership of the returned value. Use `delete [] value`

File: examples/qtmegachatapi/callGui.cpp [hunk 2]
```diff
@@ -1,70 +1,108 @@
...
+    QChar letter = (std::strlen(title) == 0)
+        ? QChar('?')
+        : QChar(title[0]);
+
+    drawAvatar(image, letter, peerHandle);
+}
+
+void CallGui::drawOwnAvatar(QImage &image)
+{   
+    const char *myName = mChatWindow->mMegaChatApi->getMyFirstname();
```
Comment (sergiohs84) [inline examples/qtmegachatapi/callGui.cpp]: You take the ownership of the returned value

File: examples/qtmegachatapi/callGui.cpp [hunk 3]
```diff
@@ -79,117 +117,135 @@ void CallGui::drawAvatar(QImage& image, QChar letter, uint64_t userid)
...
-        {
-            if (wptr.deleted())
-                return;
-            window->createCallGui(&mCall);
-            delete this;
-        });
-    }
+    KR_LOG_ERROR("=============LocalMediaFail: %s", err.c_str());
 }
+
```
Comment (sergiohs84) [inline examples/qtmegachatapi/callGui.cpp]: Spare empty lines

File: examples/qtmegachatapi/callListener.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ QImage *CallListener::CreateFrame(int width, int height, char *buffer, size_t size)
+ {
+     if((width == 0) || (height == 0))
+     {
+         return NULL;
+     }
+
+     char *copyBuff = new char[size];
+     memcpy(copyBuff,buffer,size);
+     unsigned char *auxBuf = reinterpret_cast<unsigned char*> (copyBuff);
```
Comment (sergiohs84) [inline examples/qtmegachatapi/callListener.cpp]: I think you can directly define `copyBuff` as `unsigned char*`. `memcpy()` won't complain (or you can use casting there) :)

File: examples/qtmegachatapi/chatItemWidget.cpp [hunk 5]
```diff
@@ -177,11 +177,26 @@ ChatWindow *ChatItemWidget::showChatWindow()
     return mChatWindow;
 }
 
+bool ChatItemWidget::isChatOpened()
+{
+    return (mChatWindow ?true :false);
```
Comment (sergiohs84) [inline examples/qtmegachatapi/chatItemWidget.cpp]: Style correction: `mChatWindow ? true : false`
But actually you can directly do: `return mChatWindow;` The implicity casting will take place and it's 100% safe. A pointer to null evaluates as false

File: examples/qtmegachatapi/chatSettings.cpp [hunk 6]
```diff
@@ -6,10 +6,47 @@ ChatSettings::ChatSettings(QMainWindow *parent)
     :QDialog(parent),
     ui(new Ui::SettingsDialog)
 {
+    mMainWin = (MainWindow *) parent;
     ui->setupUi(this);
+    #ifndef KARERE_DISABLE_WEBRTC
+        mega::MegaStringList *audioInDevices = mMainWin->mMegaChatApi->getChatAudioInDevices();
+        for (int i=0; i<audioInDevices->size(); i++)
```
Comment (sergiohs84) [inline examples/qtmegachatapi/chatSettings.cpp]: Style correction: `for (int i = 0; i < audioInDevices->size(); i++)`

File: examples/qtmegachatapi/chatSettings.cpp [hunk 7]
```diff
@@ -6,10 +6,47 @@ ChatSettings::ChatSettings(QMainWindow *parent)
...
 }
+
+void ChatSettings::on_buttonBox_clicked(QAbstractButton *button)
+{
+    #ifndef KARERE_DISABLE_WEBRTC
+        if (ui->audioInCombo->currentIndex() != mAudioInIdx)
+        {
+            mAudioInIdx = ui->audioInCombo->currentIndex();
+            std::string device =  ui->audioInCombo->itemText(ui->audioInCombo->currentIndex()).toLatin1().data();
+            mMainWin->mMegaChatApi->setChatAudioInDevice(device.c_str());
```
Comment (sergiohs84) [inline examples/qtmegachatapi/chatSettings.cpp]: Maybe we should add a check here. If `setChatAudioInDevice()` returns false, it means it has failed

File: examples/qtmegachatapi/localCallListener.cpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+{
+    mMegaChatApi->addChatLocalVideoListener(megaChatVideoListenerDelegate);    
+}
+
+LocalCallListener::~LocalCallListener()
+{
+    mMegaChatApi->removeChatLocalVideoListener(megaChatVideoListenerDelegate);    
+}
+
+
```
Comment (sergiohs84) [inline examples/qtmegachatapi/localCallListener.cpp]: Spare empty line

File: bindings/qt/QTMegaChatVideoListener.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+using namespace megachat;
+
+QTMegaChatVideoListener::QTMegaChatVideoListener(MegaChatApi *megaChatApi, MegaChatVideoListener *listener) : QObject()
+{
+    this->megaChatApi = megaChatApi;
+    this->listener = listener;
+}
+
+QTMegaChatVideoListener::~QTMegaChatVideoListener()
+{ }
```
[the remark this answers is not in the document]
Comment (jgandres) [reply bindings/qt/QTMegaChatVideoListener.cpp]: Yo lo estaba liberando justo después de pasárselo al callback:

char * auxBuff = event->getBuffer();
                listener->onChatVideoData(event->getMegaChatApi(), event->getChatHandle(), event->getWidth(), event->getHeight(), auxBuff, event->getSize());
                delete auxBuff;

**Comment by polmr** [comment]
trigger compilation

**Comment by polmr** [comment]
trigger compilation

**Pull request merged**